src/Firestore/ValueMapper.php

```diff
@@ -0,0 +1,429 @@
+        }
+
+        return $output;
+    }
+
+    /**
+     * Create a list of fields paths from field data.
+     *
+     * The return value of this method does not include the field values. It
+     * merely provides a list of field paths which were included in the input.
+     *
+     * @param array $fields A list of fields to map as paths.
+     * @param string $parentPath The parent path (used internally).
+     * @return array
+     */
+    public function encodeFieldPaths(array $fields, $parentPath = '')
+    {
+        $output = [];
+
+        foreach ($fields as $key => $val) {
+            if (is_array($val) && $this->isAssoc($val)) {
+                $nestedParentPath = $parentPath
+                    ? $parentPath . '.' . $key
+                    : $key;
+
+                $output = array_merge($output, $this->encodeFieldPaths($val, $nestedParentPath));
+            } else {
+                $output[] = $parentPath
+                    ? $parentPath . '.' . $key
+                    : $key;
+            }
+        }
+
+        return $output;
+    }
+
+    /**
+     * Decode field paths and values.
+     *
+     * Accepts an array of form array<string, mixed>, where the key is a field
+     * path and the value is a document field value, and returns a nested array.
+     *
+     * @param array $fieldPaths A list of field paths and values.
+     * @return array
+     */
+    public function decodeFieldPaths(array $fieldPaths)
+    {
+        $output = [];
+
+        foreach ($fieldPaths as $fieldPath => $fieldValue) {
+            $keys = explode('.', $fieldPath);
+            $num = count($keys);
+
+            $val = $fieldValue;
+            foreach (array_reverse($keys) as $index => $key) {
+                if ($num >= $index+1) {
+                    $val = [$key => $val];
+                }
+            }
+
+            $output = array_merge_recursive($output, $val);
+        }
+
+        return $output;
+    }
+
+    /**
+     * Search an array for sentinel values, returning the array of fields with
+     * sentinels removed, and a list of delete and server timestamp value field
+     * paths.
+     *
+     * @param array $fields The input field data.
+     * @return array `[$fields, $timestamps, $deletes]`
+     */
+    public function findSentinels(array $fields)
+    {
+        $timestamps = [];
+        $deletes = [];
+        $fields = $this->removeSentinel($fields, $timestamps, $deletes);
+
+        return [$fields, $timestamps, $deletes];
+    }
+
+    /**
+     * Recurse through fields and find and remove sentinel values.
+     *
+     * @param array $fields The input field data.
+     * @param array $timestamps The timestamps field paths. (reference)
+     * @param array $deletes the deletes field paths. (reference)
+     * @param string $path The current field path.
+     * @return array
+     */
+    private function removeSentinel(array $fields, array &$timestamps, array &$deletes, $path = '')
```

Comment:
This is not correctly named if it is also used to detect timestamps.

Reply:
I struggled with the naming of this method. It is intended to search the fields to find any sentinels (delete or timestamps), remove them from the fields, and track where they came from. It'll provide the list of timestamps and deletes, which can then be used to build the field mask (for deletes) or create the transforms (in the case of timestamps).